lib/babe/babe.go

```diff
@@ -355,89 +355,109 @@ func (b *Service) initiate(epoch uint64) {
 }
 
 func (b *Service) invokeBlockAuthoring(epoch uint64) {
-	// calculate current slot
-	startSlot := getCurrentSlot(b.slotDuration)
+	for {
+		// get start slot for current epoch
+		epochStart, err := b.epochState.GetStartSlotForEpoch(epoch)
+		if err != nil {
+			logger.Error("failed to get start slot for current epoch", "epoch", epoch, "error", err)
+			return
+		}
 
-	head, err := b.blockState.BestBlockHeader()
-	if err != nil {
-		logger.Error("failed to get best block header", "error", err)
-		return
-	}
+		head, err := b.blockState.BestBlockHeader()
+		if err != nil {
+			logger.Error("failed to get best block header", "error", err)
+			return
+		}
 
-	// if we're at genesis, set the first slot number for the network
-	if head.Number.Cmp(big.NewInt(0)) == 0 {
-		err = b.epochState.SetFirstSlot(startSlot)
+		// if we're at genesis, set the first slot number for the network
+		if head.Number.Cmp(big.NewInt(0)) == 0 {
+			epochStart = getCurrentSlot(b.slotDuration)
+			err = b.epochState.SetFirstSlot(epochStart)
+			if err != nil {
+				logger.Error("failed to set first slot number", "error", err)
+				return
+			}
+		}
+
+		logger.Info("initiating epoch", "number", epoch, "first slot of epoch", epochStart)
+		err = b.initiateEpoch(epoch)
 		if err != nil {
-			logger.Error("failed to set first slot number", "error", err)
+			logger.Error("failed to initiate epoch", "epoch", epoch, "error", err)
 			return
 		}
-	}
 
-	logger.Info("initiating epoch", "number", epoch, "start slot", startSlot+b.epochLength)
-	err = b.initiateEpoch(epoch)
-	if err != nil {
-		logger.Error("failed to initiate epoch", "epoch", epoch, "error", err)
-		return
-	}
+		epochStartTime := getSlotStartTime(epochStart, b.slotDuration)
+		logger.Debug("checking if epoch started", "epoch start", epochStartTime, "now", time.Now())
+
+		// check if it's time to start the epoch yet. if not, wait until it is
+		if time.Since(epochStartTime) < 0 {
+			logger.Debug("waiting for epoch to start")
+			select {
+			case <-time.After(time.Until(epochStartTime)):
+			case <-b.ctx.Done():
+				return
+			case <-b.pause:
+				return
+			}
+		}
 
-	// get start slot for current epoch
-	epochStart, err := b.epochState.GetStartSlotForEpoch(0)
-	if err != nil {
-		logger.Error("failed to get start slot for current epoch", "epoch", epoch, "error", err)
-		return
-	}
+		// calculate current slot
+		startSlot := getCurrentSlot(b.slotDuration)
+		intoEpoch := startSlot - epochStart
 
-	intoEpoch := startSlot - epochStart
-	logger.Info("current epoch", "epoch", epoch, "slots into epoch", intoEpoch)
+		// if the calculated amount of slots "into the epoch" is greater than the epoch length,
```

Comment:
is it possible to move this into another function? This seems like something that happens infrequently and can be tested independently?

Reply:
it would happen if the node is down for over an epoch or if it's starting up for the first time on a network that's past epoch 0, which part would you put in a separate function? just this `if` check?

Reply:
going to end up removing this in my next PR actually